File: scripts/javascript/async_version/memory_checker.js
const axios = require('axios');

async function fetchData() {
    const numRequests = 100;
    const memoryUsageBefore = process.memoryUsage().heapUsed;

    const requests = [];
    for (let i = 0; i < numRequests; i++) {
        const request = axios.get('https://habr.com').catch(error => {
            return null;
        });
        requests.push(request);
    }

    await Promise.all(requests);

    const memoryUsageAfter = process.memoryUsage().heapUsed;
    const memoryUsed = (memoryUsageAfter - memoryUsageBefore) / 1024 / 1024

    console.log(`Использовано памяти: ${memoryUsed} МБ`);
}

fetchData();

File: scripts/python/async_version/memory_checker.py
import asyncio
import ssl

import aiohttp
import certifi
import psutil


async def fetch(session, url):
    async with session.get(url) as response:
        return await response.text()


async def main():
    url = "https://habr.com"
    num_requests = 100

    process = psutil.Process()

    # uss "Unique Set Size" - уникальная для процесса память
    # которая будет освобождена после его закрытия
    memory_usage_before = process.memory_full_info().uss

    ssl_context = ssl.create_default_context(cafile=certifi.where())
    async with aiohttp.ClientSession(
        connector=aiohttp.TCPConnector(ssl=ssl_context)
    ) as session:
        tasks = [fetch(session, url) for _ in range(num_requests)]
        await asyncio.gather(*tasks)

    memory_usage_after = process.memory_full_info().uss
    memory_used = (memory_usage_after - memory_usage_before) / 1024 / 1024

    print(f"Использовано памяти: {memory_used} МБ")


if __name__ == "__main__":
    asyncio.run(main())
